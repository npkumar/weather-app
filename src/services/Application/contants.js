export const FAHRENHEIT = "fahrenheit";
export const KELVIN = "kelvin";
export const CELSIUS = "celsius";
