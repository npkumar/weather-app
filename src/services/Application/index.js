// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";

// Constants
export { FAHRENHEIT, CELSIUS, KELVIN } from "./contants";
