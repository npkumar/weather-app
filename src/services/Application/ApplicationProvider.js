import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";
import * as constants from "./contants";

// Actions
const SET_UNIT = "SET_UNIT";
const SET_INDEX = "SET_INDEX";

export const ApplicationContext = createContext(null);

const ApplicationProvider = ({ children }) => {
  const initialState = {
    unit: constants.FAHRENHEIT,
    index: 0,
  };

  function reducer(state, action) {
    switch (action.type) {
      case SET_UNIT: {
        return { ...state, unit: action.payload };
      }
      case SET_INDEX: {
        return { ...state, index: action.payload };
      }
      default:
        throw new Error();
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setUnit: (unit) => dispatch({ type: SET_UNIT, payload: unit }),
    setIndex: (index) => dispatch({ type: SET_INDEX, payload: index }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
