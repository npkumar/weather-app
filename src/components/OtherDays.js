import React from "react";
import PropTypes from "prop-types";
import WeatherCard from "./WeatherCard";
import { useApplication } from "../services/Application";
import { Grid } from "@material-ui/core";

const OtherDays = ({ daysData }) => {
  const { application } = useApplication();
  const yesterdayIndex = application.index - 1;
  const tomorrowIndex = application.index + 1;

  return (
    <Grid container spacing={1}>
      <Grid item xs={6}>
        {yesterdayIndex >= 0 && (
          <WeatherCard
            small
            date={daysData[yesterdayIndex].date}
            avgTempKelvin={daysData[yesterdayIndex].avgTempKelvin}
            avgTempFahrenheit={daysData[yesterdayIndex].avgTempFahrenheit}
            avgTempCelsius={daysData[yesterdayIndex].avgTempCelsius}
            data-testid="yesterday"
          />
        )}
      </Grid>
      <Grid item xs={6}>
        {tomorrowIndex < daysData.length && (
          <WeatherCard
            small
            date={daysData[tomorrowIndex].date}
            avgTempKelvin={daysData[tomorrowIndex].avgTempKelvin}
            avgTempFahrenheit={daysData[tomorrowIndex].avgTempFahrenheit}
            avgTempCelsius={daysData[tomorrowIndex].avgTempCelsius}
            data-testid="tomorrow"
          />
        )}
      </Grid>
    </Grid>
  );
};

OtherDays.propTypes = {
  daysData: PropTypes.arrayOf(
    PropTypes.shape({
      date: PropTypes.string.isRequired,
      avgTempKelvin: PropTypes.number.isRequired,
      avgTempFahrenheit: PropTypes.number.isRequired,
      avgTempCelsius: PropTypes.number.isRequired,
    })
  ).isRequired,
};

export default OtherDays;
