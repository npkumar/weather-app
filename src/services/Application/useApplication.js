import { useContext } from "react";
import { ApplicationContext } from "./ApplicationProvider";

const useApplication = () => useContext(ApplicationContext);

export default useApplication;
