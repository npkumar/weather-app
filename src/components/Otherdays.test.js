import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ApplicationContext } from "../services/Application/ApplicationProvider";
import OtherDays from "./OtherDays";

const value = {
  application: {
    index: 1,
  },
};

const daysData = [
  {
    date: "2020-07-18",
    avgTempKelvin: 286.45,
    avgTempFahrenheit: 55.94,
    avgTempCelsius: 13.3,
  },
  {
    date: "2020-07-19",
    avgTempKelvin: 288.37,
    avgTempFahrenheit: 59.4,
    avgTempCelsius: 15.22,
  },
  {
    date: "2020-07-20",
    avgTempKelvin: 287.6,
    avgTempFahrenheit: 58.01,
    avgTempCelsius: 14.45,
  },
];

describe("Otherdays", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <OtherDays daysData={daysData} />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display yesterday's and tomorrow's weather cards", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <OtherDays daysData={daysData} />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("yesterday")).toBeDefined();
    expect(getByTestId("tomorrow")).toBeDefined();
  });

  it("should only display tomorrow's weather card, if yesterday is unavailable", () => {
    const { queryByTestId, getByTestId } = render(
      <ApplicationContext.Provider value={{ application: { index: 0 } }}>
        <OtherDays daysData={daysData} />
      </ApplicationContext.Provider>
    );
    expect(queryByTestId("yesterday")).toBeNull();
    expect(getByTestId("tomorrow")).toBeDefined();
  });

  it("should only display yesterday's weather card, if tomorrow is unavailable", () => {
    const { queryByTestId, getByTestId } = render(
      <ApplicationContext.Provider value={{ application: { index: 2 } }}>
        <OtherDays daysData={daysData} />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("yesterday")).toBeDefined();
    expect(queryByTestId("tomorrow")).toBeNull();
  });
});
