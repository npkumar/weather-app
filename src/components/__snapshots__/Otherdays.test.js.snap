// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Otherdays renders 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiPaper-root makeStyles-root-1 MuiPaper-elevation3 MuiPaper-rounded"
        data-testid="yesterday"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-smallTitle-3 MuiTypography-body1 MuiTypography-gutterBottom"
            data-testid="date"
          >
            Jul 18, 2020
          </p>
          <p
            class="MuiTypography-root makeStyles-smallDayOfWeek-7 MuiTypography-body1 MuiTypography-gutterBottom"
            data-testid="day"
          >
            Saturday
          </p>
          <p
            class="MuiTypography-root makeStyles-smallDescription-5 MuiTypography-body1 MuiTypography-colorSecondary"
            data-testid="unit"
          >
             
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiPaper-root makeStyles-root-1 MuiPaper-elevation3 MuiPaper-rounded"
        data-testid="tomorrow"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-smallTitle-3 MuiTypography-body1 MuiTypography-gutterBottom"
            data-testid="date"
          >
            Jul 20, 2020
          </p>
          <p
            class="MuiTypography-root makeStyles-smallDayOfWeek-7 MuiTypography-body1 MuiTypography-gutterBottom"
            data-testid="day"
          >
            Monday
          </p>
          <p
            class="MuiTypography-root makeStyles-smallDescription-5 MuiTypography-body1 MuiTypography-colorSecondary"
            data-testid="unit"
          >
             
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
